ml_peg/analysis/utils/decorators.py

```diff
@@ -248,6 +249,269 @@ def plot_scatter_wrapper(*args, **kwargs) -> dict[str, Any]:
+
+                norm_x = (
+                    np.clip((ref_vals - ref_vals.min()) / (delta_x + eps), 0.0, 1.0)
+                    * (grid_size - 1)
+                ).astype(int)
+                norm_y = (
+                    np.clip((pred_vals - pred_vals.min()) / (delta_y + eps), 0.0, 1.0)
+                    * (grid_size - 1)
+                ).astype(int)
+
+                cell_points: dict[tuple[int, int], list[int]] = defaultdict(list)
+                for idx, (cx, cy) in enumerate(zip(norm_x, norm_y, strict=True)):
+                    cell_points[(int(cx), int(cy))].append(idx)
+
+                rng = np.random.default_rng(seed)
+                sampled_x: list[float] = []
+                sampled_y: list[float] = []
+                sampled_density: list[int] = []
+                for indices in cell_points.values():
+                    if len(indices) > max_points_per_cell:
+                        chosen = rng.choice(
+                            indices, size=max_points_per_cell, replace=False
+                        )
+                    else:
+                        chosen = indices
+                    density = len(indices)
+                    for idx in chosen:
+                        sampled_x.append(float(ref_vals[idx]))
+                        sampled_y.append(float(pred_vals[idx]))
+                        sampled_density.append(density)
+
+                return sampled_x, sampled_y, sampled_density
+
+            results = func(*args, **kwargs)
+            if not isinstance(results, dict):
+                raise TypeError(
+                    "Density plot decorator expects a mapping of model results."
+                )
+
+            model_names = list(results)
+            if not model_names:
+                raise ValueError("No results provided for density plot.")
+
+            global_min = np.inf
+            global_max = -np.inf
+            processed = {}
+            annotations = []
+            for model in model_names:
+                data = results[model]
+                ref_vals = np.asarray(data.get("ref", []), dtype=float)
+                pred_vals = np.asarray(data.get("pred", []), dtype=float)
+                meta = data.get("meta") or {}
+                excluded = meta.get("excluded")
+                excluded_text = str(excluded) if excluded is not None else "n/a"
+                if ref_vals.size == 0 or pred_vals.size == 0:
+                    sampled = ([], [], [])
+                else:
+                    sampled = _downsample(ref_vals, pred_vals)
+                    global_min = min(global_min, ref_vals.min(), pred_vals.min())
+                    global_max = max(global_max, ref_vals.max(), pred_vals.max())
+                mae_value = data.get("mae")
+                if mae_value is None and ref_vals.size:
+                    mae_value = float(np.mean(np.abs(ref_vals - pred_vals)))
+                mae_text = f"{mae_value:.3f}" if mae_value is not None else "n/a"
```

Comment:
I thought this was changed to be `metric`?

Do we need this, when it's already in the table?

Reply:
You have the model and the title, so that tells you the model row and column you selected? But my point was also the data key has changed - it's no longer `mae`, so this always recalculates?